File: src/nodes/shim.py
from .node import Node
from common.net import send_message
from common.quorum import QuorumHelper
import logging

logger = logging.getLogger(__name__)

class Shim(Node):
    def __init__(self, name, host, port, next, clients):
        super().__init__(name, host, port)
        self.next = next
        self.clients = clients
        self.quorum_helper = QuorumHelper(2)  # TODO: Replace hard-coded value with formula

    def start(self):
        super().start()

    def handle_message(self, payload):
        logger.debug(f"Handler called on {self.name} with payload: {payload}")

        msg_type = payload.get('type', 'request')
        if msg_type == 'response':
            # Handle response from exec - broadcast to all clients that sent the request
            # TODO: Or do we wait for a quorum, and then broadcast
            request_id = payload.get('request_id')
            response_data = payload.get('response')

            logger.info(f"{self.name}: Broadcasting response for request {request_id} "
                    f"to {len(self.clients)} clients: {self.clients}")

            for client in self.clients:
                try:
                    client_response = {
                        'type': 'response',
                        'request_id': request_id,
                        'response': response_data
                    }
                    send_message(client, '8000', client_response)
                    logger.debug(f"{self.name}: Sent response to client {client}")
                except Exception as e:
                    logger.error(f"{self.name}: Failed to send response to {client}: {e}")

            return {'status': 'response_broadcast', 'recipients': list(self.clients)}
        else:
            # Handle incoming client request - wait for quorum then forward
            request_id = payload.get('request_id')
            sender = payload.get('sender')

            if self.quorum_helper.add(request_id, sender):
                send_message(self.next, '8000', payload)
                return {'status': 'forwarded_to_mid_execs'}
            return {'status': 'waiting_for_quorum'}

File: src/nodes/shim.go
package nodes

import (
	"log"

	"aegean/common"
)

type Shim struct {
	*Node
	Next         string
	Clients      []string
	quorumHelper *common.QuorumHelper
}

func NewShim(name, host string, port int, next string, clients []string) *Shim {
	shim := &Shim{
		Node:         NewNode(name, host, port),
		Next:         next,
		Clients:      clients,
		quorumHelper: common.NewQuorumHelper(2), // TODO: Replace hard-coded value with formula
	}
	shim.Node.HandleMessage = shim.HandleMessage
	return shim
}

func (s *Shim) Start() {
	s.Node.Start()
}

func (s *Shim) HandleMessage(payload map[string]any) map[string]any {
	log.Printf("Handler called on %s with payload: %v", s.Name, payload)

	msgType, _ := payload["type"].(string)
	if msgType == "" {
		msgType = "request"
	}

	if msgType == "response" {
		requestID := payload["request_id"]
		responseData, _ := payload["response"].(map[string]any)

		// Handle response from exec - broadcast to all clients that sent the request
		// TODO: Or do we wait for a quorum, and then broadcast
		log.Printf("%s: Broadcasting response for request %v to %d clients: %v", s.Name, requestID, len(s.Clients), s.Clients)

		for _, client := range s.Clients {
			clientResponse := map[string]any{
				"type":       "response",
				"request_id": requestID,
				"response":   responseData,
			}
			if _, err := common.SendMessage(client, 8000, clientResponse); err != nil {
				log.Printf("%s: Failed to send response to %s: %v", s.Name, client, err)
				continue
			}
			log.Printf("%s: Sent response to client %s", s.Name, client)
		}

		return map[string]any{"status": "response_broadcast", "recipients": s.Clients}
	}

	// Handle incoming client request - wait for quorum then forward
	requestID := payload["request_id"]
	sender, _ := payload["sender"].(string)

	if s.quorumHelper.Add(requestID, sender) {
		_, _ = common.SendMessage(s.Next, 8000, payload)
		return map[string]any{"status": "forwarded_to_mid_execs"}
	}
	return map[string]any{"status": "waiting_for_quorum"}
}
